runtime/common/src/crowdloan.rs

```diff
@@ -482,23 +491,30 @@ decl_module! {
 				Error::<T>::InRetirementPeriod
 			);
 
-			let account = Self::fund_account_id(index);
+			// Try killing the crowdloan child trie
+			match Self::crowdloan_kill(index) {
+				child::KillOutcome::AllRemoved => {
+					let account = Self::fund_account_id(index);
 
-			// Avoid using transfer to ensure we don't pay any fees.
-			let transfer = WithdrawReasons::TRANSFER;
-			let imbalance = T::Currency::withdraw(&account, fund.deposit, transfer, AllowDeath)?;
-			let _ = T::Currency::resolve_into_existing(&fund.owner, imbalance);
+					// Avoid using transfer to ensure we don't pay any fees.
+					let transfer = WithdrawReasons::TRANSFER;
+					let imbalance = T::Currency::withdraw(&account, fund.deposit, transfer, AllowDeath)?;
+					let _ = T::Currency::resolve_into_existing(&fund.owner, imbalance);
 
-			let imbalance = T::Currency::withdraw(&account, fund.raised, transfer, AllowDeath)?;
-			T::OrphanedFunds::on_unbalanced(imbalance);
+					let imbalance = T::Currency::withdraw(&account, fund.raised, transfer, AllowDeath)?;
+					T::OrphanedFunds::on_unbalanced(imbalance);
 
-			Self::crowdfund_kill(index);
-			<Funds<T>>::remove(index);
+					<Funds<T>>::remove(index);
 
-			Self::deposit_event(RawEvent::Dissolved(index));
+					Self::deposit_event(RawEvent::Dissolved(index));
+				},
+				child::KillOutcome::SomeRemaining => {
+					Self::deposit_event(RawEvent::PartiallyDissolved(index));
+				}
+			}
 		}
 
-		fn on_finalize(n: T::BlockNumber) {
+		fn on_initialize(n: T::BlockNumber) -> frame_support::weights::Weight {
```

Comment:
it is hard to me to say if moving this on_initialize can be an issue, I don't know how intricated slots and crowdfunds are.

Reply:
all tests pass :)

it should only affect 1 block difference with any auction/crowdfund timelines so not a big deal.

Reply:
it might be an issue... care should be taken regarding the state transitioning of the loans. in general i'd prefer to ensure that any time the current block number is used within this call graph, it is first decremented by 1, which should make it exactly equivalent to before.

Reply:
I have checked the logic here again, and I do not see issue with the change here from `on_finalize` to `on_initialize`.

I don't think it would be right to manipulate the block number in the `on_initialize` as you hinted. Functions like `<slots::Module<T>>::is_ending(n)` would report wrong information when given the wrong block number. For example, if we tell it `is_ending(n-1)` it might report true, when `is_ending(n)` would be false because the ending period is already finished in the current block.

The only different I see between `on_initialize` vs `on_finalize` is that crowdloans do not consider donations that come in on that last block for bidding on that block. But it would increment the bid in the next block no problem.

Reply:
What was the incentive to move to on_inititalize in the first place?

Reply:
You need to move logic which may have variable/unknown weight into the beginning of the block where possible. You don't want to be left with more logic than you can fit at the end of the block. In general, `on_finalize` should only be small pieces of known size logic, not something like this crowdloan management.